src/js/components/Popover.jsx

```diff
@@ -0,0 +1,536 @@
+//
+//--------------------------------------------------------------------------
+
+export class Popover extends Component {
+
+    //--------------------------------------
+    //  Internal State
+    //--------------------------------------
+
+    lifecycleState:LifecycleState = stateInit;
+    currentLeft:number = 0;
+    currentTop:number = 0;
+
+    //--------------------------------------
+    //  Animation Progress
+    //--------------------------------------
+
+    animationStart:number = 0; // ms
+    animationDuration:number = 0; // ms
+    goalLeft:number = 0; // Goal-point for animation
+    goalTop:number = 0; // Goal-point for animation
+
+    //--------------------------------------
+    //  Last-known measurements
+    //--------------------------------------
+
+    selfWidth:number = 0;
+    selfHeight:number = 0;
+
+    targetWidth:number = 0;
+    targetHeight:number = 0;
+    targetLeft:number = 0;
+    targetTop:number = 0;
+
+    viewportWidth:number = 0;
+    viewportHeight:number = 0;
+
+    //--------------------------------------
+    //  Positioning Loop
+    //--------------------------------------
+
+    positioningValid:boolean = false;
+    validatePositioningScheduled:boolean = false;
+    checkDOMDependenciesScheduled:boolean = false;
+    pollingTimeout:any = null;
+
+    /**
+     Checks the measurements of things we depend on, and invalidate positioning
+     if they've changed. This is designed to be called often, from a timer or
+     event listeners.
+
+     Can be called at any time, and will schedule itself into an animation
+     frame.
+     */
+    checkDOMDependencies() {
+        if (this.positioningValid
+            && !this.validatePositioningScheduled
+            && !this.checkDOMDependenciesScheduled) {
+
+            this.checkDOMDependenciesScheduled = true;
+            window.requestAnimationFrame(() => {
+                this.checkDOMDependenciesScheduled = false;
+                if (this.measureDOMNodes()) {
+                    this.invalidatePositioning();
+                }
+            });
+        }
+    }
+
+    /**
+     Call when you want to make sure to re-calculate the positioning.
+
+     Can be called at any time, and will schedule itself into an animation
+     frame.
+     */
+    invalidatePositioning() {
+        if (this.positioningValid) {
+            this.positioningValid = false;
+        }
+
+        if (!this.validatePositioningScheduled) {
+            this.validatePositioningScheduled = true;
+
+            window.requestAnimationFrame(() => {
+                this.validatePositioningScheduled = false;
+                this.validatePositioning();
+            });
+        }
+    }
+
+    /**
+     Check for invalidated positioning, and re-calculate / move if necessary
+
+     Should only be called via window.requestAnimationFrame().
+     */
+    validatePositioning() {
+        if (!this.positioningValid) {
+            this.calculateAndSetPopupPosition();
```

Comment:
Should this instead change the position to what it thinks is more appropriate (vs just translating)?